계산기 [STEP 1] Rowan

Tediousday93 opened the pull request:
@Groot-94 
🌲 그루트 안녕하세요 🫡
계산기 step01 PR 보내드립니다. 
이번 프로젝트간 피드백 잘 부탁드립니다~! 🥰

## UML
![step1_UML](https://user-images.githubusercontent.com/114981173/214902026-cb6c561d-306a-49df-879e-c24b69b79834.png)


# 고민한 점
### TDD
- 처음으로 TDD를 직접 해보면서 실패 케이스 작성이 힘들었습니다.
- 무작정 메서드를 작성할 때 보다 메서드의 기능 파악을 정확히 할 수 있었던 것 같습니다.
- 테스트는 메서드의 기능을 검증하는 과정이므로 `sut` 인스턴스 생성 시, 임의로 Generic에 `Int`타입을 사용했습니다.
```swift
final class CalculatorItemQueueTests: XCTestCase {

    var sut: CalculatorItemQueue<Int>!
    
    override func setUpWithError() throws {
        var queueList = LinkedList<Int>()
        sut = CalculatorItemQueue<Int>(queue: queueList)
    }

    override func tearDownWithError() throws {
        sut = nil
    }
    ...
}
```

### Generic 사용
- Queue를 구현하다 보니 CalculatorItemQueue에서 다루는 Item들은 `CalculateItem`프로토콜을 준수한다는 조건을 보고 Queue를 여러 데이터 타입에 사용할 수 있어야 한다고 생각했습니다. 따라서 `Node` 데이터로 사용할 데이터 타입을 하나로 정해주기 보다는 Generic을 사용해 데이터에 여러 타입이 들어갈 수 있도록 나타냈습니다.

### `CalculateItem` 프로토콜
- 해당 프로토콜은 채택 시 필수 조건이 없는 빈 프로토콜로, 왜 조건에 주어졌는지 생각해봤습니다.
- CalculatorItemQueue에 전달되는 데이터의 타입을 `CalculateItem`으로 한정하면, 해당 프로토콜을 채택한 데이터 타입만 전달할 수 있는 점에서 Generic Constraint로 사용해봤습니다.
- Generic 자리에 들어갈 데이터 타입에 `extension`으로 `CalculateItem` 프로토콜을 준수하도록 만들었습니다.

```swift
struct CalculatorItemQueue<Element: CalculateItem>

extension Int: CalculateItem { }
```

### Queue List 선택
✏️ **Queue 타입 구현을 위한 List 타입**

Queue는 데이터를 List에 추가/제거하기 위한 `enqueue(_:)`와 `dequeue()` 메서드를 반드시 갖고 있어야합니다. 이때 List타입에 따라서 enqueue를 호출할 때, dequeue를 호출할 때 시간복잡도가 O(n)이 되는 문제가 있었습니다. 이를 해결하기 위해 다양한 Queue 구현 방법을 조사해보고 적절한 방법을 선택하여 구현할 수 있도록 고민했습니다.

<br/>

---

1. **Linked List**
Linked List는 떨어진 공간에 있는 데이터(Node)를 연결한 것이다. 연결 리스트를 통해 배열을 대체할 수 있다. 연결 리스트의 장단점은 아래와 같다.

![](https://i.imgur.com/GTa6FPn.png)

  * **장점**
       - 데이터 삽입, 삭제 속도가 빠르다.

<br/>

  * **단점**
      - 자료 탐색 속도가 느리다.
      - 저장 공간 효율이 좋지 않다.

    장점으로는 데이터 삽입/삭제 속도가 빠른 것이 있는데, 실질적으로 도움이 되는지는 다시 한 번 생각해봐야 할 것 같다.

    자료 탐색 속도가 느린 이유는 특정 자료에 접근하기 위해선 List가 가지고 있는 `head`라는 프로퍼티를 통해 제일 처음 Node부터 차례대로 주소를 찾아가야하기 때문이다. 또한 연결 정보를 저장하는 별도의 데이터 공간이 필요하므로 저장 공간 효율이 높지 않다.

    Queue 구현에 적용했을 때 어느정도의 시간복잡도를 갖는지 생각해보자.

    Queue의 list에 적용했을 때 필요한 기능은 앞서 말했듯이 가장 뒤쪽에 자료를 추가하는 enqueue, 제일 처음 자료를 추출하는 dequeue 메서드이다.

    dequeue의 경우, `head`를 알고 있기 때문에 선입선출(FIFO)이 용이하다. 시간복잡도는 O(1)이라고 생각한다. 반면에 enqueue의 경우 head에서 부터 가장 마지막 Node의 주소를 찾아 해당 Node에 다음 자료를 연결해주어야 하므로 시간복잡도는 O(n)이 될 것이다.

    이를 보완하기 위해 연결리스트에 `tail` 프로퍼티를 추가할 수 있다. `tail` 프로퍼티는 항상 가장 마지막 Node를 가리키고 있기 때문에 enqueue의 시간복잡도가 O(1)이 될 것이다.

    자료의 탐색에 시간이 오래 걸리고 저장 공간 효율이 좋지 않다는 단점이 있지만 이번 프로젝트의 queue 구현을 위한 List로 사용할 때, 자료의 탐색은 필요가 없을 것 같다. 따라서 Linked List를 사용하는 것도 좋은 방법이 될 것이다.

<br/>

2. **Array**
Swift에서 기본적으로 제공하는 Array를 queue의 List로 사용하면 `append(_:)` 메서드를 통해 간단하게 enqueue를 구현할 수 있다. 그러나 연결 리스트와는 다르게 dequeue에서 시간복잡도가 O(n)이 된다. 왜냐하면 Array는 Index를 갖고 있기 때문에 가장 앞쪽 데이터가 삭제되면 나머지 데이터들의 Index를 한 칸씩 이동해줘야 하기 때문이다.

    dequeue의 시간복잡도가 O(n)이 되는 문제를 해결하기 위해 개선된 Queue의 구현 방법을 알아보도록 하자.

<br/>

🧐 **Queue에 head 프로퍼티 추가하기**
    Index를 옮기는 작업을 최소화하기 위해 Queue에 head 프로퍼티를 추가해 `dequeue()`를 통해 삭제된 데이터 앞 Index를 나타내도록 한다. 데이터의 삭제는 nil을 할당하는 것으로 대체하고, nil 할당 후 head를 다음 Index로 옮겨가며 head의 Index가 특정 기준에 도달할 경우 앞쪽에 할당된 nil을 전부 삭제한다. 이 방법을 통해 `dequeue()` 메서드가 호출된 순간의 시간복잡도는 O(1)로 개선할 수 있다. 그러나 이 방법은 결국 nil을 삭제하는 과정에서 시간복잡도 O(n)을 갖게 된다.

<br/>

🧐 **Queue를 더블 스택으로 구현하기**
    후입선출(LIFO)의 구조로 데이터를 쌓아올리는 스택을 활용하여 Queue를 구현하면 dequeue의 시간복잡도 문제를 해결할 수 있다. enqueue와 dequeue를 모두 스택 구조로 만들어주었을 때 enqueue와 dequeue의 구동 방식을 살펴보자. `enqueue(_:)` 메서드를 호출할 때, enqueueStack에 데이터를 append해준다. 이후 dequeue를 진행할 때, LIFO를 이용하기 위해 enqueueStack을 뒤집어 dequeueStack에 할당해준다. Stack 내부 자료들의 순서를 뒤집는 것은 `reversed()` 메서드를 활용한다. 뒤집어진 순서의 dequeueStack에서 `popLast()` 메서드로 가장 먼저 append된 데이터를 추출한다. dequeueStack에서는 가장 마지막 데이터가 삭제되기 때문에 dequeue의 시간복잡도는 O(1)이 될 것이다.

---

검색엔진을 통해 Queue에 대한 구현 방법을 알아볼 때 일반적으로 많이 나오는 Queue는 Array를 List로 사용한 Queue들이었고, 이는 대부분 dequeue 과정에서 시간복잡도의 문제를 가지고 있었습니다. 따라서 해당 문제를 해결하기 위해 LinkedList 또는 Double Stack Queue를 선택하여 구현하는 것이 좋을 것 같습니다.

이번 프로젝트 선택사항으로 List를 직접 구현해보기도 있고, 자료 탐색이 필요없다고 생각하기 때문에 LinkedList로 구현해보았습니다.

Tediousday93 committed:
feat: CalculatorItemQueue 파일 생성

Tediousday93 committed:
feat: CalculatorItemQueueTests 파일 생성 및 최초 실패 케이스 적용

Tediousday93 committed:
feat: enqueue() 메서드 첫 테스트케이스 통과

Tediousday93 committed:
feat: dequeue메서드 구현, Test Case 3개

Tediousday93 committed:
feat: LinkedList파일, LinkedListTests파일 추가

Tediousday93 committed:
feat: appendLast()메서드 테스트케이스 2개 추가

Tediousday93 committed:
feat: appendLast()메서드 테스트케이스 2개 추가, appendList 작성완료

Tediousday93 committed:
feat: removeFirst메서드 테스트케이스 4개 추가, removeFirst 작성완료

Tediousday93 committed:
refactor: Test Case 메서드 분리를 위한 코드 수정

Tediousday93 committed:
feat: removeAll 테스트 케이스 2개 추가, removeAll 작성 완료

Tediousday93 committed:
feat: CalculatorItemQueue enqueue, dequeue, removeAll 메서드 작성

dequeue() 메서드의 반환값을 얻기 위해 LinkedList removeFirst refactoring

Tediousday93 committed:
feat: CalculateItem protocol 생성, 불필요한import foundation 삭제

Tediousday93 committed:
feat: CalculatorItemQueue의 제네릭 constraint로 CalculateItem protocol 추가

Tediousday93 committed:
feat: UML 이미지 첨부

Tediousday93 committed:
Revert "docs: UML 이미지 첨부"

This reverts commit d0fc7ff3b3ccf4429605d0877e4bbd322e8831f0.

Groot-94 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,25 @@
+//  Created by Rowan on 2023/01/24.
+//
+
+struct CalculatorItemQueue<Element: CalculateItem> {
```
요구사항에 있는 CalculateItem을 잘 파악하셨네요 👍👍👍

Groot-94 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,25 @@
+
+struct CalculatorItemQueue<Element: CalculateItem> {
+    
+    var queue: LinkedList<Element>
```
외부에서 queue에 접근해야 할 이유가 있을까요?

Tediousday93 replied:
전혀 없습니다!! 🫢

디스코드에서 답변해주신 내용을 기반으로 코드 전반에 걸쳐 외부 접근이 불필요한 프로퍼티/메서드에 대해 `private`키워드, 읽기전용 프로퍼티에 대해 `private(set)`키워드를 붙여주었습니다. 👍

Groot-94 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,61 @@
+//  Created by Rowan on 2023/01/25.
+//
+
+class Node<T> {
```
Node는 상속을 하게 될 가능성이 있을까요?

Tediousday93 replied:
상속 가능성이 없다고 생각합니다. `final` 키워드로 상속을 방지해주었습니다 👍

Groot-94 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,61 @@
+        self.data = data
+        self.next = next
+    }
+    
```
이 공백은 의도하신 공백일까요? 🤔

Tediousday93 replied:
넵 타입 정의부분 첫 줄과 맨 마지막 줄은 한 줄 공백을 주고 있습니다!
별로 보기 좋지 않은 컨벤션이라면 삭제하도록 하겠습니다 🤔

Groot-94 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,61 @@
+    
+    var head: Node<T>?
+    var tail: Node<T>?
+    var count: Int = 0
```
리스트 데이터 변경에 따라 count를 직접 수정하신 이유가 있을까요? 🤔
이 방법이 아니라면 어떻게 count를 알 수 있을까요?

Tediousday93 replied:
test case에서 count를 직접 설정해주기 보다는 sut 인스턴스 initializer를 통해 주입해주는 형태로 수정했습니다!
`count`는 `private(set)`으로 읽기전용으로 구현해 밖에서 접근만 가능하도록 정의했습니다.

나머지 `head`와 `tail`은 `private` 접근수준으로 설정하여 외부에서 두 프로퍼티 내부 data를 접근할 수 있도록 연산 프로퍼티 `headData`, `tailData`를 정의했습니다.

```swift
struct LinkedList<T> {
    
    private var head: Node<T>?
    private var tail: Node<T>?
    private(set) var count: Int
    
    var headData: T? {
        return self.head?.data
    }
    
    var tailData: T? {
        return self.tail?.data
    }
    ...
}
```

Tediousday93 replied:
질문하신 내용과 제 답변이 포인트가 맞지 않는 것 같아 하나 더 코멘트 남겨봅니당..

데이터 변경에 따라 `count`를 알게 되려면 연산 프로퍼티로 정의해줘야 할 것 같습니다.
count에 접근하여 get 블록 코드가 실행될 경우, head에서 tail까지 차례대로 .next프로퍼티를 통해 이동하면서 전체 Node의 개수를 세어주는 방법이 있을 것 같습니다.

Groot-94 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,61 @@
+    mutating func appendLast(_ data: T) {
+        let newNode = Node(data: data)
+        
+        if self.isEmpty {
```
self를 붙여주신 이유가 궁금합니다.

Tediousday93 replied:
if문의 문맥 상 self.isEmpty로 적어주는 것이 더 자연스럽게 읽힌다고 생각하여 붙여주었습니다.

Groot-94 commented on Calculator/Calculator/Model/LinkedList.swift:
```diff
@@ -0,0 +1,61 @@
+    
+    mutating func removeAll() {
+        head = nil
+        tail = nil
```
head와 tail 각각 nil을 할당해주는 방법과 
head에만 nil을 할당해주는 방법은 어떤 차이가 있을까요?🤔

Tediousday93 replied:
head에만 nil을 할당할 경우 맨 마지막 tail이 가리키고 있는 Node의 인스턴스 reference count가 0이 되지 않아 완전히 비워지지 않는 경우가 발생합니다.

따라서 완전히 데이터를 없애기 위해서는 tail에도 nil을 할당해 주어야 한다고 생각합니다.

![스크린샷 2023-01-27 오전 2 54 23](https://user-images.githubusercontent.com/114981173/214912476-d522115d-5182-4be7-862d-d5c10bb18f2e.png)

Groot-94 replied:
저는 tail노드는 항상 마지막 노드여서 tail이 가리키는 노드는 없을거라 예상되는데, 그렇게 되면 head에만 nil을 할당해줘도 ARC에 의해 다른 노드들도 메모리에서 사라진다고 생각이 됩니다. 
이 부분은 직접 테스트 해보셔도 좋을 것 같습니다.

Tediousday93 replied:
마지막 노드의 next는 가리키고 있는 노드가 없지만
tail 프로퍼티가 마지막 노드의 인스턴스를 가리키고 있기 때문에 할당 해제가 되지 않는 것을 확인했습니다!!
![스크린샷 2023-01-27 오후 6 35 06](https://user-images.githubusercontent.com/114981173/215054342-c1d43f3a-a9a4-4a1f-b379-2efb33e6e92d.png)

Groot-94 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,106 @@
+    }
+    
+//MARK: - dequeue Test
+    func test_dequeue호출시_queue의removeFirst메서드를호출한다() {
```
네이밍으로 작성해주신 removeFirst메서드 라는 부분이 함수 내부의 코드에서는 볼 수 없네요.
"removeFirst메서드" 키워드는 현재 테스트 하시려는 dequeue와는 무관해 보입니다. 
"removeFirst"는 내부에 있는 List의 테스트 케이스 같네요.

현재 테스트를 하고싶으신 기능에 초점을 맞추는게 좋을것 같아요.

Tediousday93 replied:
맞습니다... 메서드를 호출한다는 네이밍보다는 기능적으로 queue의 itemCount가 1감소한다 라고 바꿔주는게 좋아보이네용👍

Groot-94 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,106 @@
+        XCTAssertEqual(resultTail, resultExpectation)
+    }
+
+    func test_dequeue호출시_제거된element를return한다() {
```
dequeue 메서드의 동작을 상세하게 글로 잘 표현해주셨네요. 가독성 👍

Groot-94 commented on Calculator/CalculatorTests/CalculatorItemQueueTests.swift:
```diff
@@ -0,0 +1,106 @@
+    }
+    
+//MARK: - removeAll Test
+    func test_removeAll호출시_queue의removeAll이호출된다() {
```
이 부분도 위에서 말씀드린 내부함수 내용과 동일합니다.

Tediousday93 replied:
"queue의 itemCount가 0이된다" 로 수정했습니다!

Groot-94 commented:
안녕하세요 :)
저도 잘 부탁드립니다. 😆

전체적인 코드와 TDD를 나타내주신 commit 메세지 모두 인상깊었습니다. 👍👍👍👍 
아쉬운 점은 은닉화와 네이밍 부분이였습니다. 이 부분은 리팩토링에서 전반적으로 수정해주시면 좋을 것 같습니다. 

### 고민한 점 
#### TDD
- 정말 고민을 많이 하셨다는게 느껴질 정도로 Unit Test를 정말 꼼꼼하게 해주셔서 감탄했습니다. 모든 기능을 세분화해서 테스트 해주셨네요 
- 특히 함수 내부의 흐름까지 테스트하신 부분에서 대단하다고 생각했습니다. 다만, 추후에는 프로젝트 일정에 맞게 어떤 부분까지 테스트 할 지 조율하면서 작성해주시면 될 것 같습니다. 👍👍👍
- 코멘트에도 남겼지만, 테스트 시 함수의 기능자체에 초점을 맞추고 작성하시면 더 좋을 것 같습니다.
#### Generic 사용
- 이 부분은 개인적인 메세지로 말씀드렸다고 생각합니다.
#### CalculateItem 프로토콜
- 저는 개인적으로 Rowan이 프로토콜을 사용하라는 의미를 잘 파악하셨다고 느꼈습니다. 👍
#### Queue List 
- 고민하신 내용을 잘 작성해주셔서 트러블 슈팅을 보는 것 같았습니다. 추후에 이 내용을 조금 더 간결하게 정리해서 트러블 슈팅으로 작성하셔도 좋을 것 같습니다.

Tediousday93 commented:
겨우 리팩토링 하고 싶었던 부분을 다 처리했습니다아~~
그루트의 세심한 조언에 감사드립니다 🙏

private 접근제어수준 적용 후 test case 에서 발생하던 접근 수준 오류는 initializer를 통한 주입으로 해결해봤습니다.
외부에서 값에 접근하고 싶을 때도 읽기전용, 연산 프로퍼티를 구현하여 현재 제가 할 수 있는 최대한의 해결 방안을 찾아본 것 같습니다. 🧐

제네릭 사용을 고민하고 있던 부분도 잘 설명해주셔서 제가 처음에 제네릭을 사용하려고 생각했던 이유가 타당했음을 알 수 있었어요!!

이번 기회에 TDD를 요령 피우지 않고 제대로 해보고 싶어서 test case를 상세하게 적으려고 노력했던 것 같네요.
늦은 시간에도 도와주셔서 다시 한 번 감사드립니다!! 🥰

```swift
// 수정 후 test case
func test_dequeue호출시_queue의_itemCount가_1감소한다() {
        // given
        let firstNode = Node(data: 1)
        let secondNode = Node(data: 2)
        let queueList = LinkedList<Int>(head: firstNode,
                                        tail: secondNode,
                                        count: 2)
        sut = CalculatorItemQueue(queueList: queueList)
        let expectedCount = 1

        // when
        sut.dequeue()
        let count = sut.itemCount

        // then
        XCTAssertEqual(count, expectedCount)
    }
```

Tediousday93 committed:
refactor: PR feedback 적용(model 객체들의 프로퍼티에 private 접근수준 적용)

접근제어 refactoring에 따른 오류 발생으로 test case 전면 수정
LinkedList<T> 타입 appendLast 메서드 기능 분리 -> generateNode(for:) 메서드 추가, headData/tailData 연산 프로퍼티 정의
CalculatorItemQueue<Element> 타입  itemCount 연산 프로퍼티 정의, isEmpty() 메서드 정의
각 struct의 인스턴스 프로퍼티 private 접근 수준 적용으로 인한 custom initializer 정의

Tediousday93 committed:
docs: upload UML Image for Step01

Tediousday93 committed:
docs: README.md 수정

Tediousday93 committed:
docs: README.md 수정

Groot-94 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -7,12 +7,23 @@
 struct CalculatorItemQueue<Element: CalculateItem> {
     
-    var queue: LinkedList<Element>
+    private var queue: LinkedList<Element>
```
이 부분의 네이밍은 queue가 아닌 list가 맞을 것 같네요

Tediousday93 replied:
수정완료했습니당 👍

Groot-94 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -7,12 +7,23 @@
         queue.appendLast(element)
     }
     
+    @discardableResult
```
@discardableResult 는 어떤 이유로 사용되나요? 🤔

Tediousday93 replied:
반환 값이 있는 함수를 호출했을 때, 반환값이 상수나 변수에 담아지지 않았을 경우 노란색 경고문이 뜹니다!
[Result of call to 'dequeue()' is unused]
Test case에서 경고를 지워주기 위해 붙여봤습니다 👀

Groot-94 commented on Calculator/Calculator/Model/CalculatorItemQueue.swift:
```diff
@@ -22,4 +33,9 @@ struct CalculatorItemQueue<Element: CalculateItem> {
     }
+    
+    init(queueList: LinkedList<Element>) {
+        self.queue = queueList
```
init의 매개변수 네이밍과 실제 프로퍼티 네이밍을 동일하게 해주는게 더 자연스러워 보입니다...!

Tediousday93 replied:
그렇네요!! list로 수정하겠습니다 ㅎㅎ

Groot-94 commented:
수고하셨습니다. 👍
